Update report generator to implement classes

DJensen94 opened the pull request:
## 🗣 Description ##
Updated report_generator.py and pages.py to access metrics via the metrics classes previously implemented

## 💭 Motivation and context ##
This will simplify the code to limit the amount of variables that need to be passed throughout the different files in the code

## 🧪 Testing ##
We will test using the precheck tests and verify that all the reports generate correctly using the new method

## ✅ Pre-approval checklist ##

- [ X] This PR has an informative and human-readable title.
- [ X] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

DJensen94 committed:
Update pages and report generator

Update report generator to pull from the metrics classes

DJensen94 committed:
Fix class method calls

Update the method calls to match up with the new classes

DJensen94 committed:
Remove unnecessary comments

Remove unnecessary comments from reporot generator

DJensen94 committed:
fix version import

fix the version import to point to the correct location

DJensen94 committed:
fix import statements in report_generator.py

Fix import statements to be based out of the right directory

DJensen94 committed:
Add pdf generator requirements

update setup.py to include the package that creates pdfs from html

DJensen94 committed:
add importlib to setup requirements

add importlib library to the requirements list in setup

DJensen94 committed:
Fix library import

added missing comma in library import list

DJensen94 committed:
Fixed importlib install

changed importlib to importlib-resources

DJensen94 committed:
add os and change path declaration

add os library to library import and remove importlib

DJensen94 committed:
remove os install

remove install os

DJensen94 committed:
Fix path argument

Fix database.ini path definition

DJensen94 committed:
fix test imports

fix unnecessary import of outdates file

DJensen94 committed:
change section name

update change section name to match parser parameter

DJensen94 committed:
Fix local import call

fix local import to charts

DJensen94 committed:
Fixed db_query import

fixed to show correct parent directory

DJensen94 committed:
Remove calls to old file

remove calls to old report_generator file

aloftus23 committed:
Impement chart functions in Pages.py

Add pie, stacked bar, line, and horizontal bar chart functions to pages.py

aloftus23 committed:
Add logging setup to report_gnerator.py

add logging setup to main()

aloftus23 committed:
Remove docopt function call

Docopt call was causing an Attribute error.

aloftus23 committed:
Add return value

Add return to pass assertion

aloftus23 committed:
Fix logging assertion error

Args weren't saved in a Dict

aloftus23 committed:
Change return to sys.exit

return is registering as exit code

aloftus23 committed:
Create __init__

aloftus23 committed:
Fix report_generator and db_query errors.

- remove db-creds-file argument and add the log-level argument
- clean up logging in db_query and report generator
- update readme

aloftus23 committed:
Fix all errors when running the module and updated alerts

- added alerts_sites table
- fixed location of charts images
- fixed small typos

aloftus23 committed:
Fix lgtm errors

- fixed unused for loop variable in pages.py

aloftus23 committed:
Add backgroun images and remove inpur_data cmd arg

- don't need input data since it's reaching the db
- add background images for report template

aloftus23 committed:
Add CSS file and update tests

- add html css file to the repo
- update tests to remove input file

aloftus23 committed:
Fix lgtm main() issue

- have main() function return 1 if error or 0 if passes

aloftus23 committed:
Update tests to accept new return value

- tests now understand main() failing will return 1. else 0

aloftus23 committed:
Fix error in test_pe_reports.py

DJensen94 commented:
@dav3r, just wanted to check to see if you had any questions or comments

dav3r commented:
@DJensen94 I apologize for the delay - I have been occupied with other urgent priorities and have not had a chance to review these changes.  I will try to take a look at some point this week.

DJensen94 commented:
Thats great, thanks

cduhn17 commented:
@dav3r, 
Checking in to get an idea of when you might have a time slot to take a look at this PR and other outstanding PR's?
Thanks

dav3r commented on src/pe_reports/data/__init__.py:
```diff
@@ -0,0 +1 @@
+"""Initialize database folder for tests to pass."""
```
I don't understand this comment. Are you saying that the tests weren't passing unless that comment was present?

aloftus23 replied:
Updated to """The pe-reports data directory."""  [here](https://github.com/cisagov/pe-reports/pull/114/commits/6e727fe0b430b8bd02de8f3e5ae7f27c778f1925) Just a bad comment. The __init__ is so we can treat the directory as a package.

*think the "tests" we were referring to was the flask pre-commit check

dav3r replied:
Got it, thanks!

BTW, the [current linting error](https://github.com/cisagov/pe-reports/runs/4927222432?check_suite_focus=true) should get resolved once you pull in the latest code from the base branch.

dav3r commented on src/pe_reports/data/db_query.py:
```diff
@@ -20,10 +20,9 @@
     line_n = traceback.tb_lineno
-    logging.error(f"\npsycopg2 ERROR: {err} on line number: {line_n}")
-    logging.error(f"psycopg2 traceback: {traceback} -- type: {err_type}")
+    logging.error(f"Database connection error: {err} on line number: {line_n}")
```
If you're not using `line_n` anywhere else, I'd just get rid of it and change the logging statement to this:
```suggestion
    logging.error(f"Database connection error: {err} on line number: {traceback.tb_lineno}")
```

aloftus23 replied:
Fixed [here](https://github.com/cisagov/pe-reports/pull/114/commits/c76a555f77c6e7bc613bcf3dbbdf499361f251de)

dav3r commented on src/pe_reports/data/db_query.py:
```diff
@@ -20,10 +20,9 @@
     """Handle errors for PostgreSQL issues."""
-    err_type, traceback = sys.exc_info()
+    err_type, err_obj, traceback = sys.exc_info()
     line_n = traceback.tb_lineno
```
This line can be deleted if you accept my next suggestion.
```suggestion
```

aloftus23 replied:
Fixed [here](https://github.com/cisagov/pe-reports/pull/114/commits/c76a555f77c6e7bc613bcf3dbbdf499361f251de).

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -558,10 +554,27 @@ def dark_web_tags(self):
+    def alerts_site(self):
+        """Get alerts in invite-only markets."""
+        alerts = self.alerts
+        alerts_site = alerts[["site"]]
```
I'd consolidate these lines to get rid of the `alerts` variable:
```suggestion
        alerts_site = self.alerts[["site"]]
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/b3248fc1023126a5771b66f8f8e8d5c73048cece).

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -558,10 +554,27 @@ def dark_web_tags(self):
         """Get top CVEs."""
         top_cves = self.top_cves
         top_cve_table = top_cves[["cve_id", "summary"]]
+        top_cve_table = top_cve_table.copy()
```
Probably worth putting a comment in here explaining why you are making a copy of this table.

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/b3248fc1023126a5771b66f8f8e8d5c73048cece).

dav3r replied:
What exactly triggers that `SettingWithCopyWarning`?  Can you please add a little bit more explanation in that comment?

aloftus23 replied:
While looking into this more, I realized this is probably a better way to do it. No warning and is easier to understand what is going on. Let me know what you think. [Commit link](https://github.com/cisagov/pe-reports/pull/114/commits/a74c4cd4ab8df13e7e90aa0891cc3e7fbd16ae93)

dav3r replied:
That definitely seems cleaner- good thinking! 👍

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -80,7 +255,7 @@ def main() -> None:
         # Exit because one or more of the arguments were invalid
         print(err, file=sys.stderr)
-        sys.exit(1)
+        return 1
```
What's your reason for switching from `sys.exit` to `return` here?

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/a77ca5bc0ebd5cb2caf07ed12ae0e0871b00b3c9).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -90,27 +265,22 @@ def main() -> None:
+
+
+if __name__ == "__main__":
+    sys.exit(main())
```
Why was this added?  It should not be needed.

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/a77ca5bc0ebd5cb2caf07ed12ae0e0871b00b3c9).

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -1,213 +1,313 @@
+
+# Style and build tables
+def buildTable(df, classList, sizingList=[]):
+    """Build html tables from a pandas dataframe."""
```
Minor nitpick:
```suggestion
    """Build HTML tables from a pandas dataframe."""
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/a6226dc399178094a6c6f9fcd5d068d611859d5e).

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -1,213 +1,313 @@
+    pw_creds = Credential.password()
+    ce_date_df = Credential.by_days()
+    breach_det_df = Credential.breach_details()
+    breach_appendix = Credential.breach_appendix()
```
I think all of these variables are only used once below.  If that's the case, I'd get rid of them and simply use `Credential.<whatever>` where needed below.

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/a6226dc399178094a6c6f9fcd5d068d611859d5e).

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -1,213 +1,313 @@
+            "domain_table": domain_table,
+            "suspectedDomains": domain_count,
+            "uniqueTlds": utlds,
+        }
```
Similar to my previous comment, I'd remove these single-use variables:
```suggestion
    chevron_dict.update(
        {
            "domain_table": buildTable(Domain_Masq.summary(), ["table"], []),
            "suspectedDomains": Domain_Masq.count(),
            "uniqueTlds": Domain_Masq.utlds(),
        }
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/a6226dc399178094a6c6f9fcd5d068d611859d5e).

dav3r replied:
`domain_table` is still a single-use variable and can be eliminated via:
```
"domain_table": buildTable(Domain_Masq.summary(), ["table"], [])
```

DJensen94 replied:
@dav3r addressed in the following commit: [3228942](https://github.com/cisagov/pe-reports/pull/114/commits/32289424226c4801893f4e99c040e807a406713a)

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -1,213 +1,313 @@
+        verif_vulns_summary, ["table"], [15, 15, 15, 55]
+    )
+
+    # Update chevrion dictionary
```
Typo:
```suggestion
    # Update chevron dictionary
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/a6226dc399178094a6c6f9fcd5d068d611859d5e).

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -1,213 +1,313 @@
+    verif_vulns = Malware_Vuln.verif_vulns()
+    verif_vulns_summary = Malware_Vuln.verif_vulns_summary()
+    total_verif_vulns = Malware_Vuln.total_verif_vulns()
+    unverified_vuln_count = Malware_Vuln.unverified_vuln_count()
```
Some of these are single-use variables and should be removed.

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/a6226dc399178094a6c6f9fcd5d068d611859d5e).

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -1,213 +1,313 @@
+    alerts_exec_table = buildTable(alert_exec[:8], ["table"], [15, 70, 15])
+    dark_web_act_table = buildTable(dark_web_most_act, ["table"], [75, 25])
+    alerts_site_table = buildTable(alerts_site, ["table"], [50, 50])
+    top_cves_table = buildTable(top_cve_table, ["table"], [30, 70])
```
There are a lot of magic numbers in here.  Can you please add some explanatory comments?

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/82e1864bcadcc388b28b84ed41537ca6fd8b4bfc). I explained the 3rd argument in the buildTable function instead of the call here.

dav3r replied:
That works, thanks.

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -1,213 +1,313 @@
+    alert_exec = Cyber6.alerts_exec()
+    dark_web_most_act = Cyber6.dark_web_most_act()
+    alerts_site = Cyber6.alerts_site()
+    top_cve_table = Cyber6.top_cve_table()
```
More single-use variables that can be removed.

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/a6226dc399178094a6c6f9fcd5d068d611859d5e).

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -1,213 +1,313 @@
+    """Call each page of the report."""
+    # Format start_date and end_date
+    end_date = datetime.strptime(datestring, "%Y-%m-%d").date()
+    if end_date.day == 15:
```
A comment should be added explaining the significance of the 15th day of the month.

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/82e1864bcadcc388b28b84ed41537ca6fd8b4bfc).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -1,23 +1,16 @@
-"""A tool for creating Posture & Exposure reports.
+"""ciagov/pe-reports: A tool for creating Posture & Exposure reports.
```
Typo:
```suggestion
"""cisagov/pe-reports: A tool for creating Posture & Exposure reports.
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+    vuln_xlsx,
+    mi_xlsx,
+):
+    """Embeds raw data into pdf and encrypts file."""
```
Another capitalization nitpick:
```suggestion
    """Embeds raw data into PDF and encrypts file."""
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
-    """Load PowerPoint template into memory."""
-    prs = Presentation(REPORT_SHELL)
-    return prs
+    # Open csv data as binary
```
```suggestion
    # Open CSV data as binary
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
-        prs.save(os.path.join(out_dir, REPORT_OUT))
-    except FileNotFoundError as not_found:
-        logging.error("%s : Missing input data. No report generated.", not_found)
+    # Embedd and add push-pin graphic
```
Typo:
```suggestion
    # Embed and add push-pin graphic
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+    ma = open(vuln_xlsx, "rb").read()
+    mi = open(mi_xlsx, "rb").read()
 
+    # Insert link to csv data in summary page of pdf
```
```suggestion
    # Insert link to CSV data in summary page of PDF
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+):
+    """Embeds raw data into pdf and encrypts file."""
+    doc = fitz.open(file)
+    page = doc[3]
```
This magic number 3 could use an explanatory comment.

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/2599135785a1ec460be4ab00e7e33db08b7f1028).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+    mi = open(mi_xlsx, "rb").read()
 
+    # Insert link to csv data in summary page of pdf
+    p1 = fitz.Point(110, 695)
+    p2 = fitz.Point(240, 695)
+    p3 = fitz.Point(375, 695)
+    p5 = fitz.Point(500, 695)
```
More magic numbers here.

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/2599135785a1ec460be4ab00e7e33db08b7f1028).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+    )
+    page.add_file_annot(p3, ma, "vuln_alerts.xlsx", desc="Open up xlsx", icon="PushPin")
+    page.add_file_annot(
+        p5, mi, "mention_incidents.xlsx", desc="Open up csv", icon="PushPin"
```
```suggestion
        p1, cc, "compromised_credentials.xlsx", desc="Open up CSV", icon="PushPin"
    )
    page.add_file_annot(
        p2, da, "domain_alerts.xlsx", desc="Open up CSV", icon="PushPin"
    )
    page.add_file_annot(p3, ma, "vuln_alerts.xlsx", desc="Open up xlsx", icon="PushPin")
    page.add_file_annot(
        p5, mi, "mention_incidents.xlsx", desc="Open up CSV", icon="PushPin"
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+
+    doc.save(
+        output,
+        garbage=4,
```
Please add a comment here explaining `garbage=4`.

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/2599135785a1ec460be4ab00e7e33db08b7f1028).

dav3r replied:
Excellent comment, thanks!

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+    # convert HTML to PDF
+    pisa_status = pisa.CreatePDF(
+        source_html, dest=result_file  # the HTML to convert
+    )  # file handle to recieve result
```
Typo:
```suggestion
    )  # file handle to receive result
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+            org_code = org[2]
+
+            if org_code != "DOS":
+                continue
```
This definitely needs an explanatory comment. :)

aloftus23 replied:
Ah yes that shouldn't be in there. Omitted [here](https://github.com/cisagov/pe-reports/pull/114/commits/2599135785a1ec460be4ab00e7e33db08b7f1028).

dav3r replied:
Thanks- I was about to have a lot of questions about that one!

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+                os.mkdir(f"{output_directory}/ppt")
+
+            if not os.path.exists(f"{output_directory}/{org_code}"):
+                os.mkdir(f"{output_directory}/{org_code}")
```
This code can be DRY'd out a bit:
```suggestion
            for dir_name in ("ppt", org_code):
                if not os.path.exists(f"{output_directory}/{dir_name}"):
                    os.mkdir(f"{output_directory}/{dir_name}")
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/2599135785a1ec460be4ab00e7e33db08b7f1028).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+            if not os.path.exists(f"{output_directory}/{org_code}"):
+                os.mkdir(f"{output_directory}/{org_code}")
 
+            # Load source html
```
```suggestion
            # Load source HTML
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+                source_html = file.read().replace("\n", " ")
+            except FileNotFoundError:
+                logging.error(
+                    "Template cannot be found. Must be named pe-reports/src/pe_reports/template.html"
```
```suggestion
                    "Template cannot be found. It must be named: '%s'", template
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+
+
+def convert_html_to_pdf(source_html, output_filename):
+    """Convert html to pdf."""
```
```suggestion
    """Convert HTML to PDF."""
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r replied:
This change was not done in https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd - I think you overlooked it.

DJensen94 replied:
@dav3r addressed issue in the following commit: [d97b155](https://github.com/cisagov/pe-reports/pull/114/commits/d97b155c38271a4f01b868f06c89d77dadf4cfe7)

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
 
-def generate_reports(data, data_dir, out_dir, db_creds_file):
-    """Gather assets to produce reports."""
+            # Insert Charts and Metrics into pdf
```
```suggestion
            # Insert charts and metrics into PDF
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+                org_uid,
+            )
 
+            # Close pdf
```
```suggestion
            # Close PDF
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+            output_filename = f"{output_directory}/{org_code}-Posture_and_Exposure_Report-{datestring}.pdf"
+            convert_html_to_pdf(source_html, output_filename)
+
+            # Create Crendential Exposure excel file
```
```suggestion
            # Create Crendential Exposure Excel file
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+            )
+            credWriter.save()
+
+            # Create Domain Masquerading excel file
```
```suggestion
            # Create Domain Masquerading Excel file
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+            masq_df.to_excel(domWriter, sheet_name="Suspected Domains", index=False)
+            domWriter.save()
+
+            # Create Suspected vulnerability excel file
```
```suggestion
            # Create Suspected vulnerability Excel file
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+            vulns_df.to_excel(vulnWriter, sheet_name="Verified Vulns", index=False)
+            vulnWriter.save()
+
+            # Create dark web excel file
```
```suggestion
            # Create dark web Excel file
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+            top_cves.to_excel(miWriter, sheet_name="Top CVEs", index=False)
+            miWriter.save()
+
+            # grab the pdf
```
```suggestion
            # grab the PDF
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+                vuln_xlsx,
+                mi_xlsx,
+            )
+            # Need to make sure Cyhy Mailer doesn't send files that are too large
```
Make this comment a bit more direct and add some context.
```suggestion
            # Log a message if the report is too large.  Our current mailer 
            # cannot send files larger than 20MB.
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd).

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,221 @@
+                    f"{org_code} is too large. File size: {filesize} Limit: 20MB"
+                )
+
+            generated_reports = generated_reports + 1
```
Use simpler syntax:
```suggestion
            generated_reports += 1
```

aloftus23 replied:
Addressed [here](https://github.com/cisagov/pe-reports/pull/114/commits/2599135785a1ec460be4ab00e7e33db08b7f1028).

aloftus23 committed:
Update docstring in __init__ data file

Docstring in __init__ file incorrectly described the __init__ file

dav3r reviewed (changes requested):
Many apologies for taking so long to get to this review.  I had a few suggestions, questions, and bunch of minor capitalization qualms- please take a look and let me know if you have any questions.

aloftus23 committed:
Fix config file type in setup.py and remove unnecessary variables in db_query.py

- Simplify line # traceback message
- Adjusted setup.py to include data/*.ini since the cnfiguration file was changed to that file type

aloftus23 committed:
Clean up metrics.py

- remove unnecessary variable assignments
- add comment

aloftus23 committed:
Merge branch 'develop' into dj-WIP-pages

aloftus23 committed:
Change report_generator and tests to handle sys.exit()

- Instead of returning 1 on failure, sys.exit
- Change tests to handle this
- *Follows same style as pe_mailer

aloftus23 committed:
Clean pages.py comments and single-use variables

- Fix comment typos
- Remove single-use variable declarations

aloftus23 committed:
Clarify the dates and table function in pages.py

- Explain in comments how we get the date span and explain how the table widths are declared

aloftus23 committed:
Clean-up comments in report_generator.py

- capitalize acronyms and products
- clean-up spelling

aloftus23 commented:
@dav3r I have quick question regarding all the typos in report_generator.py. Would you rather we fix all of them in one commit like I did [here](https://github.com/cisagov/pe-reports/pull/114/commits/9c39af6967a25d26434abdd25b958c107b77ccbd)? OR should I have just used the "commit suggestion" button for each of them?

dav3r commented:
> @dav3r I have quick question regarding all the typos in report_generator.py. Would you rather we fix all of them in one commit like I did here? OR should I have just used the "commit suggestion" button for each of them?

What you did is fine, though personally in cases like this, I prefer to use the "Add suggestion to batch" feature in the Github web GUI and then do a single commit that includes all of the related suggestions in the batch.

aloftus23 committed:
Add additional comments and clean redundancies in report_generator.py

- Further explain numbers used in fitz module
- Remove temporary if statement (DOS)

aloftus23 commented:
> > @dav3r I have quick question regarding all the typos in report_generator.py. Would you rather we fix all of them in one commit like I did here? OR should I have just used the "commit suggestion" button for each of them?
> 
> What you did is fine, though personally in cases like this, I prefer to use the "Add suggestion to batch" feature in the Github web GUI and then do a single commit that includes all of the related suggestions in the batch.

Ok great, that's good to know. I will make sure to do that next time. Better yet, we'll do better with our spelling :)

aloftus23 commented:
@dav3r Addressed all of your suggestions. Thanks!

dav3r commented on tests/test_pe_reports.py:
```diff
@@ -84,7 +83,7 @@ def test_reports_log_levels(level):
                 logging.getLevelName(logging.root.getEffectiveLevel()) == level.upper()
             ), f"root logger level should be set to {level.upper()}"
-            assert return_code is None, "main() should return success"
+            assert return_code is None, "main() should return success (0)"
```
When you made the updates in https://github.com/cisagov/pe-reports/pull/114/commits/a77ca5bc0ebd5cb2caf07ed12ae0e0871b00b3c9, you missed this:
```suggestion
            assert return_code is None, "main() should return success"
```

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,223 @@
+):
+    """Embeds raw data into PDF and encrypts file."""
+    doc = fitz.open(file)
+    # Get the the summary page of the PDF on page 4
```
Remove duplicated word:
```suggestion
    # Get the summary page of the PDF on page 4
```

aloftus23 committed:
Fix test_report and report_generator typos

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,223 @@
+            output_filename = f"{output_directory}/{org_code}-Posture_and_Exposure_Report-{datestring}.pdf"
+            convert_html_to_pdf(source_html, output_filename)
+
+            # Create Crendential Exposure Excel file
```
Just noticed this typo:
```suggestion
            # Create Credential Exposure Excel file
```

dav3r commented on src/pe_reports/data/db_query.py:
```diff
@@ -20,10 +20,10 @@
-    logging.error(f"\npsycopg2 ERROR: {err} on line number: {line_n}")
-    logging.error(f"psycopg2 traceback: {traceback} -- type: {err_type}")
+    err_type, err_obj, traceback = sys.exc_info()
+    logging.error(
+        f"Database connection error: {err} on line number: {traceback.tb_lineno}"
+    )
```
One thing that I overlooked during my initial review is that we no longer use f-strings in our logging messages- [this post](https://www.siegescape.com/blog/2019/10/3/be-careful-with-your-f-strings) is a good short read on why we do that.

Please go through all of your logging statements and ensure that there are no f-strings in them.  If there are, you should either fix them now (preferable) or create an issue to fix them in the future.

```suggestion
    logging.error(
        "Database connection error: %s on line number: %s", err, traceback.tb_lineno
    )
```

DJensen94 replied:
@dav3r addressed f-strings in the following commit: [5ad2c0b](https://github.com/cisagov/pe-reports/pull/114/commits/5ad2c0bf0ab8cf21b2d9adfb69c9158ac9f0cebd)

dav3r replied:
Those look great now, aside from one minor thing that I'm about to note as a separate item.

DJensen94 committed:
Update src/pe_reports/report_generator.py

commit suggested change of fixed typo on comment

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Remove f-strings from all logging

Remove all the f-strings from logging statements

DJensen94 committed:
Remove single use variable from pages

Remove single use variable for domain_table

DJensen94 committed:
Capitalize acronyms in comments

Capitalize HTML and PDF

aloftus23 committed:
Fix top_cve table clean-up in metrics.py

- instead of making a deep copy of the df to silence the warning, we decided to make a new column so that there was no confusion

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -28,39 +21,223 @@
+
+    # Iterate over organizations
+    if pe_orgs:
+        logging.info("PE orgs count: %s", len(pe_orgs))
```
Not a big deal, but since we know this one is a number:
```suggestion
        logging.info("PE orgs count: %d", len(pe_orgs))
```

aloftus23 committed:
Change % type to digit when logging numbers

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r reviewed (approved):
Thanks for taking care of everything I noted - this all looks good to me now! 💎

cduhn17 commented on src/pe_reports/data/config.py:
```diff
@@ -2,9 +2,10 @@
 # Standard Python Libraries
 from configparser import ConfigParser
-from importlib.resources import files
+import os
```
@DJensen94 
If in the event that we wanted to use importlib, it would be`from  importlib_resources import files`. I remember the problems we had when this import requirement changed. 
@dav3r 
Do you have a recommendation as to which method is preferred, when reading in the `database.ini` file?

dav3r replied:
I think it's best to use `importlib` if possible here.  Is there a reason that you moved away from it?

cduhn17 replied:
@dav3r 
If i remember correctly there was some deprecation that took place in the importlib package and its a new package but implemented the same. Ok thanks for your input. We can get this changed back to using importlib.

dav3r replied:
Are you maybe thinking of [this conversation](https://github.com/cisagov/pe-reports/pull/91#discussion_r739378886), where I recommended switching from `pkg_resources` to `importlib`?

DJensen94 replied:
@dave, @cduhn17 issue addressed in the following commits [997c197](https://github.com/cisagov/pe-reports/pull/114/commits/997c197be90ee91fc5d24e635bc7c2c74fe82ee2), [9272975](https://github.com/cisagov/pe-reports/pull/114/commits/9272975ae126ab0bca525ba362493393ea22a6ac), [64b2290](https://github.com/cisagov/pe-reports/pull/114/commits/64b22908e1c9ed9b01799241a00a3a9761a41d78)

cduhn17 replied:
@DJensen94 , 

Looks good and thanks for making those changes!

cduhn17 reviewed (approved):
See comments made at config.py. The outcome in using one method over the other doesn't change functionality currently. If there is a preference we can change in the next PR.

DJensen94 committed:
reimplement importlib instead of os

use import lib to access files location

DJensen94 committed:
update importlib import

adjust the importlib statement to updated form

DJensen94 committed:
update filepath of database config

update file location of db config to correct file
